astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -8,14 +8,138 @@
 
 import numpy as np
 
-from ... units import dimensionless_unscaled, UnitsError
-from ... import log
-from ...extern.six import string_types
-from ..nduncertainty import IncompatibleUncertaintiesException
+from ..nduncertainty import NDUncertainty
+from ...units import dimensionless_unscaled
+from ...extern import six
 
 __all__ = ['NDArithmeticMixin']
 
 
+# TODO: General question of implementing an Operator class #3861? (embray)
+# Maybe a bit too big for these changes.
+
+try:
+    from .utils import format_doc
+except ImportError:
+    # TODO: Delete this and rebase if #4242 is merged.
+    def format_doc(docstring, *args, **kwargs):
+        def set_docstring(func):
+            if not isinstance(docstring, six.string_types):
+                doc = docstring.__doc__
+            elif docstring != 'self':
+                doc = docstring
+            else:
+                doc = func.__doc__
+                func.__doc__ = None
+            if not doc:
+                raise ValueError
+            kwargs['original_doc'] = func.__doc__ or ''
+            func.__doc__ = doc.format(*args, **kwargs)
+            return func
+        return set_docstring
+
+# TODO: Maybe not nice to pollute globals but I felt the same way about
+# polluting the class namespace and this way potential subclasses may or refuse
+# to pull this docstring into their class as well.
+
+# Docstring templates for add, subtract, multiply, divide methods.
+_arit_doc = """
+    Performs {name} by evaluating ``self`` {op} ``operand``.
+
+    For the inverse operation or operations between different subclasses take a
+    look at
+    :meth:`~astropy.nddata.NDArithmeticMixin.ic_{name}` .
+
+    Parameters
+    ----------
+    operand : `NDData`-like instance or convertable to one.
+        The second operand in the operation.
+
+    propagate_uncertainties : `bool` or ``None``, optional
+        If ``None`` the result will have no uncertainty. If ``False`` the
+        result will have a copied version of the first operand (or the
+        second if the first had no uncertainty). In case this is ``True``
+        the result will have a correctly propagated uncertainty from the
+        uncertainties of the operands but this assumes that the uncertainties
```

Comment:
The documentation is wrong it only must be an ``NDUncertainty`` but for everything except ``StdDevUncertainty`` the ``propagate`` (or better the ``_convert_uncertainty``) raises an Error.

It is possible though to rewrite ``_arithmatics_uncertainty`` to either:
- check for ``hasattr(self.uncertainty, 'propagate')`` instead of ``isinstance(self.uncertainty, NDUncertainty)``. This makes the documentation really hard since I would need to explain: 
  1. What happens if one (and only one) of both uncertainties is None => it wraps it inside another ``other | self.uncertainty.__class__(None)`` 
  2. that the uncertainty implements a ``propagate`` method 
  3. that the propagate method must have the same signature as ``NDUncertainty.propagate`` 
  4. That the result-parameter for ``propagate`` may be an ``astropy.units.Quantity`` (actually this one isn't really necessary because it is mentioned in ``NDUncertainty_propagate``).
- Edit the documentation that it needs to be an ``NDUncertainty``-like class that implements propagation.

I'd lean to the second option and let subclasses that want to implement ``propagation`` for other uncertainties to override ``_arithmatics_uncertainty`` - With the splitting of methods this would be quite easy and probably only a one-liner. But that's just easier for me, I can see the advantages of the first option. :-)